File: Medium/1282. Group the People Given the Group Size They Belong To/1282. Group the People Given the Group Size They Belong To.js
/**
 * @param {number[]} groupSizes
 * @return {number[][]}
 */
var groupThePeople = function (groupSizes) {
    const groups = {}; // Using an object as a dictionary to store groups
    const result = []; // Initialize the result array

    for (let i = 0; i < groupSizes.length; i++) {
        const size = groupSizes[i];

        if (!groups[size]) {
            groups[size] = []; // Initialize the group if it doesn't exist
        }

        groups[size].push(i); // Add the person to the corresponding group

        if (groups[size].length === size) {
            // If the group is full, add it to the result and reset it
            result.push(groups[size]);
            groups[size] = [];
        }
    }

    for (const size in groups) {
        if (groups[size].length > 0) {
            result.push(groups[size]); // Add any remaining groups to the result
        }
    }

    return result;
};

File: Medium/1282. Group the People Given the Group Size They Belong To/1282. Group the People Given the Group Size They Belong To.py
from typing import List


class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        groups = {}  # Using a dictionary to store groups
        result = []  # Initialize the result list

        for i, size in enumerate(groupSizes):
            if size not in groups:
                groups[size] = []  # Initialize the group if it doesn't exist

            groups[size].append(i)  # Add the person to the corresponding group

            if len(groups[size]) == size:
                # If the group is full, add it to the result and reset it
                result.append(groups[size])
                groups[size] = []

        for size, group in groups.items():
            if group:
                result.append(group)  # Add any remaining groups to the result

        return result
